### PR TITLE
Media stream recorder

### PR DESCRIPTION
**Pull request by samedhi**
This allows browsers to record from the webcam into local Blobs.

### CODE REVIEW

**Comment by crisptrutski** [comment]
Hi @samedhi, thanks for the contribution!

Just gave this a spin, and ran into two roadbumps:

1. `navigation` object used to get the stream needs externs for advanced compilation. 
    @martinklepsch is there a story around externs for browser builtins. Do cljsjs packages for just externs make sense? Just worked around this with the `js*` macro for testing.

2. Using the constructor as in your suggested usage is not returning a valid recorder, it's not returning anything in fact.

```clojure
(def media-constraints #js {:audio false :video true})

(defn record! [stream]
  (prn stream) ;; stream here is a MediaStream object
  (let [msr (js/MediaStreamRecorder stream)]
    ;; msr here is nil
    (.start msr) ;; null pointer error
    ))

(js* "navigator.getUserMedia(packages_test.core.media_constraints,
                             packages_test.core.record_BANG_,
                             cljs.core.prn)")

```

This seems to just be an error in your documentation - the constructor should be called like this: `(js/MediaStreamRecorder. stream)`

Even then I'm getting an exception when calling `(.start msr)`:

```
Uncaught TypeError: Cannot read property 'indexOf' of undefined
MediaStreamRecorder.start	@	MediaStreamRecorder-standalone.inc.js:19
packages_test$core$record_BANG_	@	core.cljs:33
```

Any ideas? Just want to have given things a basic test drive before publishing :smile:

**Comment by crisptrutski** [comment]
@samedhi Poked around in the MSR docs, and constructed a working example:

```clojure 
(defn record! [stream]
  (let [msr (js/MediaStreamRecorder. stream)]
    (aset msr "mimeType" "video/webm")
    (aset msr "width" 320)
    (aset msr "height" 240)
    (set! (.-ondataavailable msr)
          #(js/console.log %))
    (.start msr 0)))

(def media-constraints #js {:audio false :video true})

(js/navigator.webkitGetUserMedia media-constraints
                                 record!
                                 js/console.warn)
```

Would you mind updating the README you provided to give the correct form for calling the constructor, and mention that `mimeType` and dimensions are necessary before recording?

Otherwise good to go :)

**Comment by samedhi** [comment]
Thanks for looking, let me know of any other problems. BTW, I tried matchbox a little while ago, pretty darn neat.

File: media-stream-recorder/resources/cljsjs/media-stream-recorder/common/media-stream-recorder.ext.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/**
+ * @fileoverview Externs for MediaStreamRecorder version 1.2.6
+ */
+
+/**
+ * @constructor
+ */
+var MultiStreamRecorder = function(stream) {};
+var MultiStreamRecorder = function(stream) {};
```
Comment (martinklepsch) [inline media-stream-recorder/resources/cljsjs/media-stream-recorder/common/media-stream-recorder.ext.js]: This seems to be duplicate?

Comment (samedhi) [inline media-stream-recorder/resources/cljsjs/media-stream-recorder/common/media-stream-recorder.ext.js]: I am amazed at the things I miss.

**Comment by crisptrutski** [comment]
@samedhi thanks for the kind mention of Matchbox! Just pushed out 0.0.7 now, the poor library has been sitting on snapshot for some time, with a total rewrite in 0.0.8 that I need to come back to and finish. Great time to add some decent documentation now that CLJS is bootstrapped. Come hack on it ;)

This PR looks good now, I'll squash and publish.

**Comment by crisptrutski** [comment]
Merged in 18d304a64b920cecf2c492cd2f88578d2c394124 and published to clojars. Thanks!

**Pull request closed without merging**